lib/recursive-methods.rb

```diff
@@ -1,49 +1,73 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - recurses for each element
+# Space complexity: O(n) - recurses for each element
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError, "Negative number" if n < 0 
+    return 1 if n == 1 || n == 0
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - recurses for each element
+# Space complexity: O(n^2) - new string for each recursion
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    return "" if s.length == 0 
+    return reverse(s[1,s.length]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    return rev_help(s, 0, s.length - 1)  
 end
+  
+def rev_help(s, first_letter, last_letter)
+    if first_letter < last_letter
+        temp = s[first_letter]
+        s[first_letter] = s[last_letter]
+        s[last_letter] = temp
 
-# Time complexity: ?
-# Space complexity: ?
+        return rev_help(s, first_letter + 1, last_letter - 1)
+    end
+
+    return s
+end
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    return 0 if n == 0
+    return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - recurses for each element
+# Space complexity: O(n^2) - new string for each recursion
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    return true if s == ""
+    return false if s.length % 2 != 0
+    return s[0] == "(" && s[-1] == ")" && nested(s[1,s.length - 2])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+    return false if array.empty?
+    element = array.shift
+    return element == value || search(array, value)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def is_palindrome(s)
```

Comment:
Since you're making a new array with each recursive call this is O(n^2) in space complexity.